python/cudf_polars/cudf_polars/experimental/io.py

```diff
@@ -257,33 +280,161 @@ def do_evaluate(
+        }
 
-    # For each column, calculate the `total_uncompressed_size` for each file
-    for name, uncompressed_sizes in metadata.columnchunk_metadata().items():
-        column_sizes[name] = np.array(
-            [
-                np.sum(uncompressed_sizes[start:end])
-                for (start, end) in itertools.pairwise(rowgroup_offsets_per_file)
-            ],
-            dtype="int64",
+        # Collect real unique-count of first row-group
+        # TODO: Check for 'distinct_count' statistics
+        # and make this sampling configurable.
+        options = plc.io.parquet.ParquetReaderOptions.builder(
+            plc.io.SourceInfo(paths[:1])
+        ).build()
+        options.set_columns([c for c in ir.schema if c in need_schema])
+        options.set_row_groups([[0]])
+        tbl_w_meta = plc.io.parquet.read_parquet(options)
+        row_group_num_rows = tbl_w_meta.tbl.num_rows()
+        unique_count_estimates: dict[str, int] = {}
+        for name, column in zip(
+            tbl_w_meta.column_names(), tbl_w_meta.columns, strict=True
+        ):
+            if name in need_schema:
+                row_group_unique_count = plc.stream_compaction.distinct_count(
+                    column,
+                    plc.types.NullPolicy.INCLUDE,
+                    plc.types.NanPolicy.NAN_IS_NULL,
+                )
+                # Assume that if every row is unique then this is a
+                # primary key otherwise it's a foreign key and we
+                # can't use the single row group count estimate
+                # Example, consider a "foreign" key that has 100
+                # unique values. If we sample from a single row group,
+                # we likely obtain a unique count of 100. But we can't
+                # necessarily deduce that that means that the unique
+                # count is 100 / num_rows_in_group * num_rows_in_file
+                if row_group_unique_count == row_group_num_rows:
+                    unique_count_estimates[name] = num_rows_total
+
+        # Construct estimated TableStats
+        table_stats = TableStats(
+            column_stats={
+                name: ColumnStats(
+                    dtype=dtype,
+                    unique_count=unique_count_estimates.get(name),
+                    element_size=element_sizes[name],
+                    file_size=total_uncompressed_size[name],
+                )
+                for name, dtype in need_schema.items()
+            },
+            num_rows=num_rows_total,
         )
 
-    # Return the mean per-file `total_uncompressed_size` for each column
-    return {name: np.mean(sizes) for name, sizes in column_sizes.items()}
+        if table_stats_cached is not None:
+            # Combine new and cached column stats
+            table_stats = TableStats.merge([table_stats_cached, table_stats])
+
+    _update_tablestats_cache(tuple(ir.paths), table_stats)
+
+    if ir.predicate is not None:
```

Comment:
Needed to move this correction until *after* we cache the `TableStats`. Otherwise we will use the wrong `row_count` the next time we read from the same list of files.